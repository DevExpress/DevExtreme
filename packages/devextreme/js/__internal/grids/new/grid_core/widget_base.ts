/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
/* eslint-disable @typescript-eslint/explicit-function-return-type */
/* eslint-disable @typescript-eslint/no-non-null-assertion */
/* eslint-disable spellcheck/spell-checker */
import browser from '@js/core/utils/browser';
import type dxScrollable from '@js/ui/scroll_view/ui.scrollable';
import { isMaterialBased } from '@js/ui/themes';
import Widget from '@js/ui/widget/ui.widget';
import { DIContext } from '@ts/core/di/index';
import type { Subscription } from '@ts/core/reactive/index';
import { ColumnsChooserView } from '@ts/grids/new/grid_core/columns_chooser/view';
import { ColumnsController } from '@ts/grids/new/grid_core/columns_controller/columns_controller';
import { DataController } from '@ts/grids/new/grid_core/data_controller/data_controller';
import { EditingController } from '@ts/grids/new/grid_core/editing/controller';
import { HeaderPanelController } from '@ts/grids/new/grid_core/header_panel/controller';
import { HeaderPanelView } from '@ts/grids/new/grid_core/header_panel/view';
import { MainView } from '@ts/grids/new/grid_core/main_view';
import { PagerView } from '@ts/grids/new/grid_core/pager';
import { Search } from '@ts/grids/new/grid_core/search/controller';
import { render } from 'inferno';

import { ContentView } from './content_view/content_view';
import { StatusView } from './content_view/status_view/status_view';
import { ErrorController } from './error_controller/error_controller';
import { FilterPanelView } from './filtering/filter_panel/filter_panel';

export class GridCoreNew<Properties> extends Widget<Properties> {
  protected renderSubscription?: Subscription;

  protected diContext!: DIContext;

  private dataController!: DataController;

  private columnsController!: ColumnsController;

  private editingController!: EditingController;

  private pagerView!: PagerView;

  private columnsChooser!: ColumnsChooserView;

  private headerPanelController!: HeaderPanelController;

  private headerPanelView!: HeaderPanelView;

  private errorController!: ErrorController;

  private search!: Search;

  protected _registerDIContext(): void {
    this.diContext = new DIContext();
    this.diContext.register(DataController);
    this.diContext.register(ColumnsController);
    this.diContext.register(HeaderPanelController);
    this.diContext.register(HeaderPanelView);
    this.diContext.register(EditingController);
    this.diContext.register(PagerView);
    this.diContext.register(ColumnsChooserView);
    this.diContext.register(Search);
    this.diContext.register(StatusView);
    this.diContext.register(FilterPanelView);
    this.diContext.register(ContentView);
    this.diContext.register(ErrorController);
  }

  protected _initDIContext(): void {
    this.columnsChooser = this.diContext.get(ColumnsChooserView);
    this.dataController = this.diContext.get(DataController);
    this.columnsController = this.diContext.get(ColumnsController);
    this.headerPanelController = this.diContext.get(HeaderPanelController);
    this.headerPanelView = this.diContext.get(HeaderPanelView);
    this.editingController = this.diContext.get(EditingController);
    this.pagerView = this.diContext.get(PagerView);
    this.search = this.diContext.get(Search);
    this.errorController = this.diContext.get(ErrorController);
  }

  protected _init(): void {
    // @ts-expect-error
    super._init();
    this._registerDIContext();
    this._initDIContext();
  }

  // eslint-disable-next-line max-len
  // eslint-disable-next-line @typescript-eslint/explicit-function-return-type, @typescript-eslint/explicit-module-boundary-types
  protected _getDefaultOptions() {
    return {
      // @ts-expect-error
      ...super._getDefaultOptions() as {},
      paging: {
        pageSize: 6,
        pageIndex: 0,
      },
      searchText: '',
      editingChanges: [],
      toolbar: {
        visible: true,
      },
    };
  }

  protected _defaultOptionsRules() {
    // @ts-expect-error
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
    return super._defaultOptionsRules().concat([
      {
        device() {
          // @ts-expect-error
          return isMaterialBased();
        },
        options: {
          headerFilter: {
            height: 315,
          },
          editing: {
            useIcons: true,
          },
          selection: {
            showCheckBoxesMode: 'always',
          },
        },
      },
      {
        device() {
          return browser.webkit;
        },
        options: {
          loadingTimeout: 30, // T344031
          loadPanel: {
            animation: {
              show: {
                easing: 'cubic-bezier(1, 0, 1, 0)',
                duration: 500,
                from: { opacity: 0 },
                to: { opacity: 1 },
              },
            },
          },
        },
      },
    ]);
  }

  protected _initMarkup(): void {
    // @ts-expect-error
    super._initMarkup();
    // @ts-expect-error
    this.renderSubscription = this.diContext.get(MainView).render(this.$element().get(0));
  }

  protected _clean(): void {
    this.renderSubscription?.unsubscribe();
    // @ts-expect-error
    render(null, this.$element().get(0));
    // @ts-expect-error
    super._clean();
  }

  public getScrollable(): dxScrollable {
    return this.diContext.get(ContentView).scrollableRef.current!;
  }
}
