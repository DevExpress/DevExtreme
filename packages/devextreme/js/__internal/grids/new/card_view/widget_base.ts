/* eslint-disable spellcheck/spell-checker */
/* eslint-disable @typescript-eslint/no-non-null-assertion */
import registerComponent from '@js/core/component_registrator';
import { Content as ContentBase } from '@ts/grids/new/grid_core/content_view/content';
import { MainView as MainViewBase } from '@ts/grids/new/grid_core/main_view';
import { OptionsController as OptionsControllerBase } from '@ts/grids/new/grid_core/options_controller/options_controller';
import { GridCoreNew } from '@ts/grids/new/grid_core/widget_base';

import { Content } from './content_view/content_view_content';
import { HeadersView } from './headers/view';
import { MainView } from './main_view';
import { OptionsController } from './options_controller';
import type { Properties } from './types';

export class CardView extends GridCoreNew<Properties> {
  protected _registerDIContext(): void {
    super._registerDIContext();
    this.diContext.register(HeadersView);

    this.diContext.register(ContentBase, Content);
    this.diContext.register(MainViewBase, MainView);

    const optionsController = new OptionsController(this);
    this.diContext.registerInstance(OptionsController, optionsController);
    // @ts-expect-error
    this.diContext.registerInstance(OptionsControllerBase, optionsController);
  }
}

// @ts-expect-error
registerComponent('dxCardView', CardView);

export default CardView;
