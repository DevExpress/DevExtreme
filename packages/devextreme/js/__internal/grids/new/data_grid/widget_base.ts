/* eslint-disable spellcheck/spell-checker */
import registerComponent from '@js/core/component_registrator';
import { Content as ContentBase } from '@ts/grids/new/grid_core/content_view/content';
import { MainView as MainViewBase } from '@ts/grids/new/grid_core/main_view';
import { OptionsController as OptionsControllerBase } from '@ts/grids/new/grid_core/options_controller/options_controller';
import { GridCoreNew } from '@ts/grids/new/grid_core/widget_base';

import { Content } from './content_view/content_view_content';
import { MainView } from './main_view';
import { OptionsController } from './options_controller';
import type { Properties } from './types';

class DataGridNew extends GridCoreNew<Properties> {
  protected _registerDIContext(): void {
    super._registerDIContext();
    this.diContext.register(ContentBase, Content);
    this.diContext.register(MainViewBase, MainView);

    const optionsController = new OptionsController(this);
    this.diContext.registerInstance(OptionsController, optionsController);
    // @ts-expect-error
    this.diContext.registerInstance(OptionsControllerBase, optionsController);
  }
}

// @ts-expect-error
registerComponent('dxDataGridNew', DataGridNew);

export default DataGridNew;
