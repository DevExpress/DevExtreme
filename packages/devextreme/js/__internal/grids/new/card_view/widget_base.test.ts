/* eslint-disable @typescript-eslint/no-unused-vars */
import { CardView } from './widget_base';

describe('common', () => {
  describe('initial render', () => {
    it('should be successfull', () => {
      const container = document.createElement('div');
      const cardView = new CardView(container, {
        dataSource: [{ a: 'a' }],
      });

      expect(container).toMatchSnapshot();
    });
  });
});

describe('options', () => {
  describe('disabled', () => {
    it('should add dx-state-disabled class to container div', () => {
      const container = document.createElement('div');
      const cardView = new CardView(container, {
        disabled: true,
      });

      expect(container.classList).toContain('dx-state-disabled');
    });
  });
});
