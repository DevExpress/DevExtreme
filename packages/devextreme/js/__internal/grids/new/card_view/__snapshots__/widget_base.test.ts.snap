// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common initial render should be successfull 1`] = `
<div
  class="dx-widget"
>
  <div
    class="dx-toolbar dx-widget dx-visibility-change-handler dx-collection"
    role="toolbar"
  >
    <div
      class="dx-toolbar-items-container"
    >
      <div
        class="dx-toolbar-before"
        role="presentation"
      />
      <div
        class="dx-toolbar-center"
        role="presentation"
      />
      <div
        class="dx-toolbar-after"
        role="presentation"
      >
        <div
          class="dx-item dx-toolbar-item dx-toolbar-button dx-toolbar-item-auto-hide"
        >
          <div
            class="dx-item-content dx-toolbar-item-content"
          >
            <input />
          </div>
        </div>
        <div
          class="dx-item dx-toolbar-item dx-toolbar-button dx-toolbar-item-auto-hide dx-toolbar-text-auto-hide"
        >
          <div
            class="dx-item-content dx-toolbar-item-content"
          >
            <div
              aria-haspopup="dialog"
              aria-label="column-chooser"
              class="dx-button dx-button-normal dx-button-mode-contained dx-widget dx-button-has-icon"
              role="button"
              tabindex="0"
            >
              <div
                class="dx-button-content"
              >
                <i
                  class="dx-icon dx-icon-column-chooser"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="dx-item dx-toolbar-item dx-toolbar-button"
        >
          <div
            class="dx-item-content dx-toolbar-item-content"
          >
            <div
              aria-label="edit"
              class="dx-button dx-button-normal dx-button-mode-contained dx-widget dx-button-has-text"
              role="button"
              tabindex="0"
            >
              <div
                class="dx-button-content"
              >
                <span
                  class="dx-button-text"
                >
                  edit
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          aria-disabled="true"
          class="dx-state-disabled dx-item dx-toolbar-item dx-toolbar-button"
        >
          <div
            class="dx-item-content dx-toolbar-item-content"
          >
            <div
              aria-label="revert"
              class="dx-button dx-button-normal dx-button-mode-contained dx-widget dx-button-has-text"
              role="button"
              tabindex="-1"
            >
              <div
                class="dx-button-content"
              >
                <span
                  class="dx-button-text"
                >
                  revert
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          aria-disabled="true"
          class="dx-state-disabled dx-item dx-toolbar-item dx-toolbar-button"
        >
          <div
            class="dx-item-content dx-toolbar-item-content"
          >
            <div
              aria-label="save"
              class="dx-button dx-button-normal dx-button-mode-contained dx-widget dx-button-has-text"
              role="button"
              tabindex="-1"
            >
              <div
                class="dx-button-content"
              >
                <span
                  class="dx-button-text"
                >
                  save
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="dx-toolbar-button dx-toolbar-menu-container dx-state-invisible"
        >
          <div
            aria-expanded="false"
            aria-haspopup="true"
            aria-label="overflow"
            class="dx-dropdownmenu dx-dropdownmenu-button dx-button dx-button-normal dx-button-mode-contained dx-widget dx-button-has-icon"
            role="button"
            tabindex="0"
          >
            <div
              class="dx-button-content"
            >
              <i
                class="dx-icon dx-icon-overflow"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="dx-gridcore-headers"
  >
    <div
      class="dx-sortable"
    >
      <div
        class="dx-gridcore-header-item"
      >
        a
      </div>
    </div>
  </div>
  <div
    class="dx-scrollable dx-visibility-change-handler dx-scrollable-vertical dx-scrollable-simulated"
  >
    <div
      class="dx-scrollable-wrapper"
    >
      <div
        class="dx-scrollable-container"
        tabindex="0"
      >
        <div
          class="dx-scrollable-content"
          style="left: 0px; top: 0px; transform: none;"
        >
          <div
            class="dx-gridcore-content"
            tabindex="0"
          >
            
            <div
              class="dx-cardview-card"
              tabindex="0"
            >
              <div
                class="dx-cardview-field"
                tabindex="0"
              >
                <span
                  class="dx-cardview-field-name"
                >
                  a
                  : 
                </span>
                <span
                  class="dx-cardview-field-name"
                  style="text-align: left;"
                >
                  a
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="dx-scrollable-scrollbar dx-widget dx-scrollbar-vertical dx-scrollbar-hoverable"
        >
          <div
            class="dx-scrollable-scroll dx-state-invisible"
            style="height: 15px; transform: translate(0px, 0px);"
          >
            <div
              class="dx-scrollable-scroll-content"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  
  
  <div>
    <div
      _feedbackhidetimeout="400"
      _feedbackshowtimeout="30"
      activestateenabled="false"
      addwidgetclass="true"
      aria="[object Object]"
      aria-label="Page Navigation"
      class="dx-widget dx-pager"
      classes="dx-pager"
      csstext=""
      focusstateenabled="false"
      hoverstateenabled="false"
      name=""
      role="navigation"
      rootelementref="[object Object]"
      rtlenabled="false"
      visible="true"
    >
      <div
        class="dx-page-sizes"
      >
        <div
          aria-label="Items per page: 2"
          class="dx-page-size dx-first-child"
          role="button"
          tabindex="0"
        >
          2
        </div>
        <div
          aria-label="Items per page: 6"
          class="dx-page-size dx-selection"
          role="button"
          tabindex="0"
        >
          6
        </div>
        <div
          aria-label="Items per page: 18"
          class="dx-page-size"
          role="button"
          tabindex="0"
        >
          18
        </div>
      </div>
      <div
        class="dx-pages"
        style="visibility: hidden;"
      >
        <div
          class="dx-page-indexes"
        >
          <div
            aria-current="page"
            aria-label="Page 1"
            class="dx-page dx-selection"
            role="button"
            tabindex="0"
          >
            1
          </div>
        </div>
      </div>
    </div>
  </div>
  
</div>
`;
